spatialos-sdk/src/worker/entity_builder.rs

```diff
@@ -0,0 +1,161 @@
+use crate::worker::{
+    component::Component,
+    component::ComponentId,
+    entity::Entity,
+    internal::schema::{
+        SchemaComponentData, SchemaDouble, SchemaObject, SchemaObjectField, SchemaPrimitiveField,
+        SchemaString, SchemaStringField, SchemaUint32,
+    },
+};
+use std::collections::HashMap;
+
+const ENTITY_ACL_COMPONENT_ID: ComponentId = 50;
+const METADATA_COMPONENT_ID: ComponentId = 53;
+const POSITION_COMPONENT_ID: ComponentId = 54;
+const PERSISTENCE_COMPONENT_ID: ComponentId = 55;
+
+pub struct EntityBuilder {
+    entity: Entity,
+
+    position: (f64, f64, f64),
+    is_persistent: bool,
+    metadata: Option<String>,
+
+    write_permissions: HashMap<ComponentId, String>,
+    read_permissions: Vec<String>,
+
+    error: Option<String>,
+}
+
+impl EntityBuilder {
+    pub fn new<T: Into<String>>(x: f64, y: f64, z: f64, position_write_layer: T) -> Self {
+        let builder = EntityBuilder {
+            entity: Entity::new(),
+            is_persistent: false,
+            metadata: None,
+            position: (x, y, z),
+            write_permissions: HashMap::new(),
+            read_permissions: Vec::new(),
+            error: None,
+        };
+
+        builder.set_write_access(POSITION_COMPONENT_ID, position_write_layer)
+    }
+
+    pub fn add_component<C: Component, T: Into<String>>(mut self, data: C, write_layer: T) -> Self {
+        if let Err(e) = self.entity.add(data) {
+            self.error = Some(e);
+        };
+
+        self.write_permissions.insert(C::ID, write_layer.into());
+        self
+    }
+
+    pub fn set_persistent<T: Into<String>>(mut self, write_layer: T) -> Self {
+        self.is_persistent = true;
+        self.set_write_access(PERSISTENCE_COMPONENT_ID, write_layer)
+    }
+
+    pub fn set_metadata<T: Into<String>, U: Into<String>>(
+        mut self,
+        entity_type: T,
+        write_layer: U,
+    ) -> Self {
+        self.metadata = Some(entity_type.into());
+        self.set_write_access(METADATA_COMPONENT_ID, write_layer)
+    }
+
+    pub fn set_read_access<T: AsRef<str>>(mut self, layers: &[T]) -> Self {
```

Comment:
I'm not sure about the way read/write access are being specified :thinking: I like that you've set the API to reflect the new layers system (e.g. it only allows a single write layer to be specified per component), but I don't know that separating the read layers from specifying the component is a good idea.

It seems like the ideal approach would be to specify the component, its write layer, and its read layer(s) all at the same time. The issue is how to do that in an ergonomic way. There are a couple of approaches I can think of:

We could have a separate builder for for components, so it would look something like:

```rust
let entity = EntityBuilder::new()
    .add(
        ComponentBuilder::new(my_component)
            .write_layer("foo")
            .read_layers(&["foo", "bar", "baz"]),
    );
```

The most obvious drawback here is that it's really verbose and annoying to use. A potentially less verbose alternative (that could be built on top of the component builder) is to use a macro:

```rust
let entity = EntityBuilder::new()
    .add(component!(
        my_component,
        write: "foo",
        read: ["foo", "bar", "baz"],
    ));
```

We could possibly extend this even further by providing a macro for the entire entity builder:

```rust
let entity = entity![
    { my_component, write: "foo", read: ["foo", "bar", "baz"] },
    { other_component, write: "bar", read: ["bar", "baz"] },
];
```

I generally avoid using macros, but in this case I think it greatly improves the ergonomics, and can just use the `EntityBuilder`/`ComponentBuilder` API under the hood.

Reply:
Read access is declared at an entity level rather than a component level :wink:

Reply:
Oh right, sorry 🤦‍♀️ Well in that case, maybe make the write layer parameter an `Option<String>`? Or even better an `Option<T> where T: Into<String>` so you can do `Some("foo")`.

I also wonder if `set_write_access` is needed at all? Since you set the write access of a component type when it's added, do we need an additional way to set write access?

Reply:
Hmm - what would the use case be for declaring that nothing can ever write to a component? 

In practice, this can be achieved by passing in a `""` argument - but I'd prefer not to encode it as part of the API. 

---
The reason I've added `set_write_access` is mainly for letting a user add an owner for the `EntityAcl` component, this could be more clearly focused to: 
`set_entity_acl_write_access`?

Reply:
I can imagine cases where you'd want to create read-only components in order to associate immutable data with an entity. I expect that some of the built-in components are often never modified (e.g. `Persistence` and `Metadata`). Though I get your point that it's probably waaaaaaaaaaaay more common to assign write access to a component, so maybe have a separate `add_immutable` method that doesn't take a write parameter? That would still allow users to add immutable components while making it more clear that they're doing so (instead of accidentally forgetting to assign a write layer).

Reply:
Yeah, I think something like `set_acl_write_access` would be clearer (and simpler, since it wouldn't need the component ID parameter) 👍